### ISSUE TITLE
Support for charm class selector

### ISSUE THREAD

**Issue by gnuoy** [canonical/operator#171]
In the OpenStack reactive charms we have put in place code which selects the charm class to use based on the version of OpenStack to be installed and the version of Ubuntu running the charm. Here is an example of the resulting charm classes [https://github.com/openstack/charm-aodh/blob/master/src/lib/charm/openstack/aodh.py#L129 ](url) 

It would be really useful if the operator framework supported something like this. The charm author could provide a selector function that the framework could use to identify the right charm class to use for the deploy.

**Comment by ajkavanagh** [comment]
Just to add to this, a little, as an idea of what could be done:

So the code in charms.openstack starting around https://github.com/openstack/charms.openstack/blob/master/charms_openstack/charm/core.py#L225 is very OpenStack dependent, so the objective here isn't to replicate that.  @gnuoy and I chatted a little, and the purpose is to try to reduce boilerplate in the charm, not necessarily replicate what we've done elsewhere.

I think the question can be boiled down to:

1. Support variations in charm behaviour due to different series of Ubuntu AND/OR different versions/releases of the payload software.
2. Try to do that without introducing lots of `if: then: else:` statements into event handling functions to dispatch behaviour based on some parameters.

The way this was solved in charms.openstack was to decide that (for the payload versions) there would be a different class for changes in behaviour.  e.g. the 'base' charm might handle versions a..c, but a derived charm (with some different behaviour) would handle version d onwards.

This might be extended/adapted to handle different versions of Ubuntu (which curiously, charms.openstack *does* handle via if-then type statements!).

So, conceptually, this could be completely orthogonal to the call of `main(DeriviedOSCharmClass)`, with some standardised helpers that:

- Allow marking of a series of derived classes in some way (e.g. via metaclass, or decorator) ... this would end up in a list of the possible classes.  The metadata approach would enable this to be transparent in the background.
- a function that can be supplied that when it is called with the list of classes gives up one that is the class to be instantiated.

So, I guess the minimum implementation, is just a mechanism to get a list of classes that can be selected between (by an arbitrary charm-specific function) so that behaviour can be modified through inheritance, and provide this as a thing in the framework.

**Comment by jameinel** [comment]
I don't quite see why that needs to be in the framework. A function that
knows about a registry of Charm classes that uses its own logic to decide
which one it wants to use, and then pass that into main() feels like it can
be done cleanly as a shared component that you just import.
I'm not sure what the framework could provide that would make it any better
than just a shared lib.
If that pattern turns out to be super useful to lots of charms, we could
pull it into the framework but it seems orthogonal.

John
=:->

**Comment by niemeyer** [comment]
Agree with @jameinel here. On the framework and juju side, we need to make sure two different possibilities are well supported:
1. Charm has independent channels providing _different_ content for each channel
2. Charm has independent channels providing the _same_ content for each channel
With those possibilities, the code can be organized either to handle both versions with the same code, or handle both versions with independent code.

In other words, handling OpenStack can be done either by the same code, or by different code, by your choice. If you do with the same code, it should be easy to have independent classes to handle different versions of OpenStack, but that's specific to the particular software being managed.

Closing this for now on that basis. Let us know if it makes sense, though.

**Comment by johnsca** [comment]
+1 to what John and Gustavo said and just wanted to add that the charm store supporting tracks seems like an even better way to organize charm code that is coupled to an application version.  E.g., the CK snaps already use tracks for the versions, and being able to use those same tracks on the charm would allow us to greatly simplify the logic in the charm code by only maintaining code relevant to the supported version.

**Comment by ajkavanagh** [comment]
Cory, just to clarify, it wasn't different versions of the charm, but the payload.  i.e. the tracks in the charm store is perfect for managing different releases of the charms, and we'll almost certainly be making use of that feature.

What this was about is the charm handling different versions of the payload.  In our case, OpenStack charms handle multiple versions of OpenStack payloads in any single version of the charm.  e.g. the current release handles OpenStack mitaka on trusty all the way through to (almost) Ussuri on Focal in the 'master' branch of the charms.  That's multiple versions of Ubuntu and multiple versions of OpenStack (as a 2d matrix of support) supported in individual charms, as it's *mostly* the same for each combination.  Obviously, it's the differences that make it tricky!

In 'classic' charms, this was handled with `if` statements, selecting between the different code paths in each function that did something for the payload.  In reactive charms the *OpenStack* versions were handled by different classes that inherited from common base classes.  A 'selector' (at run time) then chose which class to instantiate which would provide the modified behavior for that version of OpenStack.  Note that Ubuntu versions were still handled via `if` statements!

Thus, this issue was opened to see if there might be a better way of handling this 2d matrix of support in the operator framework.

However, on reflection, I think Gustavo is probably right.  It can be handled orthogonally to the *framework*, and it's probably better not to add the feature into it.

Thanks for all your thoughts/consideration on this!

**Comment by johnsca** [comment]
@ajkavanagh My point was that if you need to dispatch to a different charm class implementation for each workload version, then you are essentially maintaining multiple versions of the charm but within a single code base.  In that case, it seems like it would make more sense to directly couple the charm code to the workload version, using tracks in the charm store and release branches in your SCM.  That way, you don't have to carry forward charm code related to old versions of the workload, simplifying your charm logic and reducing the potential source of bugs (e.g., from unexpected interactions between code for new and old workload versions).
